## 4.6 (a) Combining & Exporting Data

#### This script contains the following points:
1. Import the data sets into jupiter
2. Check the dimensions of the imported dataframes
3. Merge the dataframes
4. Use the indicator argument to check if there was a match between the two dataframes
5. Export data in pickle format

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets into jupiter

In [3]:
# Tell Python to remember a main folder path
path = r'/Users/gideon/Desktop/27-06-2020 Instacart Basket analysis'

In [7]:
# Import dataset orders_products_prior.csv
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [8]:
# Import dataset orders_checked.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

### 2. Check the dimensions of the imported dataframes

In [9]:
# Check the output
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [10]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     7.0  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

In [12]:
df_ords.shape

(3421083, 8)

In [13]:
# Drop unnecessary columns
df_ords = df_ords.drop(['Unnamed: 0.1', 'Unnamed: 0'], axis = 1)

In [14]:
# Check the output
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     7.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [15]:
df_ords.shape

(3421083, 6)

In [16]:
df_ords_prior.shape

(32434489, 4)

### 3. Merge the dataframes 

In [18]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [19]:
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     7.0         196                  1          0   both  
1                     7.0       14084                  2          0   both  
2                     7.0       12427                  3          0   both  
3                     7.0       26088                  4          0   both  
4                     7.0       26405                  5          0   both

### 4. Use the indicator argument to check whether there was a full match between the two dataframes

In [21]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

#### After merging, the resulting dataframe has 32,434,489 rows, and each of these rows has information found in both input data sets, as we used an inner join for this project.

### 5. Export data in pickle format

In [22]:
# Export data to pkl
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [23]:
# Export modified df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_final.csv'))